Clarify Path.GetPathRoot return value when path is effectively empty

carlossanlop opened the pull request:
As discussed here: https://github.com/dotnet/corefx/issues/41278

stephentoub commented on xml/System.IO/Path.xml:
```diff
@@ -1459,7 +1459,7 @@ Unlike the string overload, this method doesn't normalize directory separators.
         <param name="path">The path from which to obtain root directory information.</param>
         <summary>Gets the root directory information of the specified path.</summary>
-        <returns>The root directory of <paramref name="path" />, or <see langword="null" /> if <paramref name="path" /> is <see langword="null" />, or an empty string if <paramref name="path" /> does not contain root directory information.</returns>
+        <returns>The root directory of <paramref name="path" />, or <see langword="null" /> if <paramref name="path" /> is <see langword="null" /> or only contains spaces, or an empty string if <paramref name="path" /> does not contain root directory information.</returns>
```
The "or only contains spaces" is specific to Windows, no?

JeremyKuhne replied:
Yes, that is correct. Only Windows.

carlossanlop replied:
I will make sure to clarify that as well.

carlossanlop commented:
@JeremyKuhne @stephentoub based on my findings when testing the GetPathRoot methods (both the string and ROS overloads), I modified their documentation and removed the exception mention. Can you please take a look again when you get a chance?

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+        <param name="path">A read-only span of characters containing the path from which to obtain root directory information.</param>
         <summary>Gets the root directory information from the path contained in the specified character span.</summary>
-        <returns>A character span containing the root directory of `path`.</returns>
+        <returns>A read-only span of characters containing the root directory of `path`.</returns>
```
```suggestion
        <returns>A read-only span of characters containing the root directory of <paramref name="path" />.</returns>
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+ 
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
```
The examples section should be deleted, since it illustrates the wrong overload.

carlossanlop replied:
```suggestion
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
 
 ## Remarks
 
+This method does not verify that the path or file name exists.
```
```suggestion
This method does not verify that the path or file exists.
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
 
+A `ReadOnlySpan<System.Char>` is "effectively empty" if:
+
+- In Windows, calling `IsEmpty` on this span of characters returns `true`, or all its characters are spaces (' ').
```
```suggestion
- In Windows, calling <xref:System.ReadOnlySpan%7BSystem.Char%7D.IsEmpty?displayProperty=nameWithType> on this span of characters returns `true`, or all its characters are spaces (' ').
```
I think the xref will work.

carlossanlop replied:
Cross fingers

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+- In Windows, calling `IsEmpty` on this span of characters returns `true`, or all its characters are spaces (' ').
+- In Unix, calling `IsEmpty` on this span of characters returns `true`.
+
+Possible patterns for the string returned by this method are as follows:
```
```suggestion
Possible patterns for the read-only character span returned by this method are as follows:
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+Possible patterns for the string returned by this method are as follows:
+
+- `null` (`path` was null or an empty string).
```
```suggestion
- <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType> (`path` was <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType>.
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+A `ReadOnlySpan<System.Char>` is "effectively empty" if:
+
+- In Windows, calling `IsEmpty` on this span of characters returns `true`, or all its characters are spaces (' ').
+- In Unix, calling `IsEmpty` on this span of characters returns `true`.
```
```suggestion
- In Unix, calling <xref:System.ReadOnlySpan%7BSystem.Char%7D.IsEmpty?displayProperty=nameWithType> on this span of characters returns `true`.
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+- `null` (`path` was null or an empty string).
+
+- An empty string (`path` specified a relative path on the current drive or volume).
```
```suggestion
- <xref:System.ReadOnlySpan%7BSystem.Char%7D.Empty?displayProperty=nameWithType> (`path` specified a relative path on the current drive or volume).
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+- "\" (Unix: `path` specified an absolute path on the current drive).
+
+- "X:" (Windows: `path` specified a relative path on a drive, where X represents a drive or volume letter).
```
```suggestion
- "X:" (Windows: `path` specified a relative path on a drive, where *X* represents a drive or volume letter).
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
```
```suggestion
- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions, and in .NET Framework 4.6.2 and later versions).
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
+
+## Remarks
+
+This method does not verify that the path or file name exists.
```
```suggestion
This method does not verify that the path or file exists.
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
+
+- "\" (Unix: `path` specified an absolute path on the current drive).
+
+- "X:" (Windows: `path` specified a relative path on a drive, where X represents a drive or volume letter).
```
```suggestion
- "X:" (Windows: `path` specified a relative path on a drive, where *X* represents a drive or volume letter).
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
+
+- "\\\ComputerName\SharedFolder" (Windows: a UNC path).
+
+- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions and in .NET Framework 4.6.2 and later versions)
```
```suggestion
- "\\\\\?\C:" (Windows: a DOS device path, supported in .NET Core 1.1 and later versions, and in .NET Framework 4.6.2 and later versions)
```

rpetrusha commented on xml/System.IO/Path.xml:
```diff
@@ -1457,46 +1489,58 @@ Unlike the string overload, this method doesn't normalize directory separators.
+          ]]></format>
         </remarks>
-        <exception cref="T:System.ArgumentException">
-          <paramref name="path" /> contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.  
```
Does the method not throw if `path` contains invalid characters?

carlossanlop replied:
That's what I discovered on .NET Core 3.0. Let me test this again but on Framework, I suspect this comment applies only for that version.

carlossanlop replied:
@rpetrusha confirmed:
- In 3.0, we don't throw when we pass "`\0`", `null` or `string.Empty`.
    - If we pass `null` or `string.Empty`, `GetPathRoot` returns `null`.
    - If we pass "`\0`", `GetPathRoot` returns `string.Empty`.
- In Framework, we throw if we pass "`\0`" or `string.Empty`. We don't throw if we pass `null`.

@JeremyKuhne is this a breaking change? Should I open an issue to bring back the exception in .NET Core for those cases?

carlossanlop replied:
@rpetrusha I brought back the exception, but made sure to mention it is only thrown in .NET Framework (for both root causes).

JeremyKuhne replied:
It was a deliberate breaking change in 2.1. Partially to align with the Span overloads (there are no null Spans), partially to align better with the other APIs and user expectations.

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
+[!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
+[!code-vb[System.IO.Path Members#8](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.IO.Path Members/VB/pathmembers.vb#8)]
```
```suggestion
```

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
+[!code-csharp[System.IO.Path Members#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.IO.Path Members/CS/pathmembers.cs#8)]
```
```suggestion
```

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
+[!code-cpp[System.IO.Path Members#8](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.IO.Path Members/CPP/pathmembers.cpp#8)]
```
```suggestion
```

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
+
```
```suggestion
```

carlossanlop commented on xml/System.IO/Path.xml:
```diff
@@ -1402,16 +1402,48 @@ The following example defines a variable, `basePath`, to represent an applicatio
+For more information on file paths on Windows, see [File path formats on Windows systems](~/docs/standard/io/file-path-formats.md). For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common-i-o-tasks.md).
+
+## Examples
+The following example demonstrates a use of the `GetPathRoot` method.
```
```suggestion
```

mairaw commented on xml/System.IO/Path.xml:
```diff
@@ -1457,47 +1483,64 @@ Unlike the string overload, this method doesn't normalize directory separators.
-  
- <see cref="F:System.String.Empty" /> was passed to <paramref name="path" />.</exception>
-        <related type="article" href="~/docs/standard/io/file-path-formats.md">File path formats on Windows systems</related>
+        <exception cref="T:System.ArgumentException"><paramref name="path" />.NET Framework only: contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.
```
should this be like this?

```suggestion
        <exception cref="T:System.ArgumentException">.NET Framework only: <paramref name="path" /> contains one or more of the invalid characters defined in <see cref="M:System.IO.Path.GetInvalidPathChars" />.
```

carlossanlop commented:
@mairaw I think this one's fully cooked unless you say otherwise.

carlossanlop commented:
@mairaw I addressed the comments, and Ron approved it. Can you please merge this whenever you get a chance?
